compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2774,8 +2777,26 @@ private void checkErrorConstructorInvocation(BLangInvocation iExpr) {
         setErrorReasonParam(iExpr, reasonArgGiven, ctorType);
         setErrorDetailArgsToNamedArgsList(iExpr);
 
-        resultType = expType;
-        iExpr.symbol = lhsErrorType.ctorSymbol;
+        resultType = expectedError;
+        iExpr.symbol = expectedError.ctorSymbol;
+    }
+
+    private BErrorType getExpectedErrorType(DiagnosticPos pos, BType expType, BSymbol iExprSymbol) {
+        if (expType.tag == TypeTags.UNION) {
+            List<BType> matchedErrors = ((BUnionType) expType).getMemberTypes().stream()
+                    .filter(m -> types.isAssignable(m, iExprSymbol.type))
+                    .collect(Collectors.toList());
+            if (matchedErrors.size() == 1) {
```

Comment:
IIRC spec before adding indirect-error-constructors specifically prohibited using union type as lhs when using direct error constructor. I think it's safe to simply deny this scenario.

Reply:
I think we need to support this scenario. 
```
const F = "Foo";
const G = "Foo";

type E1 error<F, record { string message?; error cause?;}>;
type E2 error<G, record { string message?; error cause?;}>;
type E E1|E2;

public function main() {
    E x = E1(); // Ok, since it say E1.
}
```